extensions/rvadapter/src/main/java/com/google/android/agera/rvadapter/RepositoryPresenterCompilerStates.java

```diff
@@ -53,12 +56,60 @@
   }
 
   /**
+   * Compiler state to optionally enable fine-grained events using {@link DiffUtil}.
+   */
+  interface RPDiff<TVal, TRet> {
+
+    /**
+     * Specifies a key extractor {@link Function} to enable generating fine-grained events using
+     * {@link DiffUtil}. Two item objects are {@linkplain Callback#areItemsTheSame the same item} if
+     * their keys are {@linkplain Object#equals equal}; and they have {@linkplain
+     * Callback#areContentsTheSame the same content} if they themselves are {@linkplain
+     * Object#equals equal}. However, if the presenter is given the same instance of collection
+     * object to generate the fine-grained events, then the presenter will consider it caused by an
+     * update from an {@linkplain Builder#addAdditionalObservable additional observable}, and send a
+     * blanket item change event covering all positions.
+     *
+     * <p>Note that {@link DiffUtil} may not be suitable for all situations due to the computation
+     * complexity and that <i>the computation is likely done from the main thread</i>. If client
+     * code has better alternatives, it should directly subclass {@link RepositoryPresenter} and
+     * provide a custom implementation in {@link RepositoryPresenter#getUpdates}.
+     *
+     * @param keyForItem A function from individual item objects to objects uniquely identifying the
+     *     items across any changes. Can be the same as the stable ID function, provided the
+     *     function does not return the same ID for multiple items (in particular
+     *     {@link RecyclerView#NO_ID}.
+     * @param detectMoves The same parameter for {@link DiffUtil#calculateDiff(Callback, boolean)}.
+     * @see RepositoryPresenter#getUpdates
+     */
+    @NonNull
+    TRet diffWith(@NonNull Function<? super TVal, ?> keyForItem, boolean detectMoves);
+  }
+
+  /**
    * Compiler state to compile for the specified item container type of the associated
    * {@link Repository}.
    */
   interface RPItemCompile<TVal> {
 
     /**
+     * Enables fine-grained events catered for the single-item special case, where the item is
+     * considered always {@linkplain Callback#areItemsTheSame the same}. The only possible events
+     * are:
+     * <ul>
+     * <li>a single item is inserted, if the old data contains no item and the new data contains the
+     *     item;
+     * <li>a single item is removed, if the old data contains the item and the new data contains no
+     *     item;
+     * <li>a single item is changed, if the old item object and the new item object are not equal,
+     *     or if the {@link RepositoryPresenter#getUpdates} call is likely due to an update from an
+     *     {@linkplain Builder#addAdditionalObservable additional observable}.
+     * </ul>
+     */
+    @NonNull
+    RPItemCompile<TVal> diff();
```

Comment:
when the single item can be a totally different item (i.e. areItemsTheSame needs to return false sometimes). In this case you'll need to do diffWith().